### PR TITLE
Adapt main_loop for regression task

### PR DESCRIPTION
**Pull request by fral92**
* Add regularization and visualization option for models that performs OF prediction
* Add dataset specific extra parameters flag
* Divide operations in build_graph, main_loop and validate depending on the type of task

### CODE REVIEW

File: main_loop_tf/config/flow.py [hunk 1]
```diff
@@ -17,6 +17,10 @@
                       lower_bound=1)
 gflags.DEFINE_bool('do_validation_only', False, 'If True does one round '
                    'of validation')
+gflags.DEFINE_string('task', 'regression',
+                     '[classification/segmentation/regression]')
```
Comment (fvisin) [inline main_loop_tf/config/flow.py]: One in [classification, segmentation, regression].

Explain the differences in detail.

File: main_loop_tf/config/optimization.py [hunk 2]
```diff
@@ -58,3 +58,16 @@
...
+gflags.DEFINE_string('of_regularization_type', None, 'Type of regularization '
+                     'to be applied to the of gradients if any. For now only '
+                     'Huber penalty is available (huber_penalty)')
+# Parameters for the optical flow regularizer (e.g. Huber penalty)
+# For the Huber penalty they are:
+# - version [custom/tf]
+# - delta
+# - weight_decay
+gflags_ext.DEFINE_multidict('of_regularization_params', {},
+                            'Optical flow regularization params')
```
Comment (fvisin) [inline main_loop_tf/config/optimization.py]: Move the comment inside the parameter description.

File: main_loop_tf/main.py [hunk 3]
```diff
@@ -144,8 +150,12 @@ def __parse_config(argv=None):
     try:
         Dataset = getattr(dataset_loaders, cfg.dataset)
     except AttributeError:
-        Dataset = getattr(dataset_loaders, cfg.dataset.capitalize() +
-                          'Dataset')
+        try:
+            Dataset = getattr(dataset_loaders, cfg.dataset.capitalize() +
+                              'Dataset')
```
Comment (fvisin) [inline main_loop_tf/main.py]: Replace with
Dataset = getattr(dataset_loaders, cfg.dataset[0].upper() + cfg.dataset[1:] + 'Dataset')

File: main_loop_tf/main.py [hunk 4]
```diff
@@ -144,8 +150,12 @@ def __parse_config(argv=None):
...
+        except AttributeError:
+            Dataset = getattr(dataset_loaders, cfg.dataset + 'Dataset')
```
Comment (fvisin) [inline main_loop_tf/main.py]: Remove this.

File: main_loop_tf/main.py [hunk 5]
```diff
@@ -185,6 +195,8 @@ def __parse_config(argv=None):
     dataset_params['divide_by_per_img_std'] = cfg.divide_by_per_img_std
     dataset_params['remove_mean'] = cfg.remove_mean
     dataset_params['divide_by_std'] = cfg.divide_by_std
+    # Add dataset extra parameters specific for each dataset
+    dataset_params.update(cfg.train_extra_params)
```
Comment (fvisin) [inline main_loop_tf/main.py]: Replace at L 160
dataset_params = {}
with
dataset_params = cfg.train_extra_params

File: main_loop_tf/main.py [hunk 6]
```diff
@@ -198,6 +210,8 @@ def __parse_config(argv=None):
         'use_threads': False,  # prevent shuffling
         # prevent crop
         'data_augm_kwargs': {'return_optical_flow': cfg.of}})
+    # Add dataset extra parameters specific for each dataset
+    cfg.valid_params.update(cfg.val_extra_params)
```
Comment (fvisin) [inline main_loop_tf/main.py:207]: Ditto

File: main_loop_tf/main.py [hunk 7]
```diff
@@ -334,6 +348,9 @@ def __run(build_model):
                                              cfg.decay_steps,
                                              cfg.decay_rate,
                                              staircase=cfg.staircase)
+        elif cfg.lr_decay == 'custom':
+            epoch = tf.cast(cfg.global_step / cfg.decay_steps, tf.int32)
```
Comment (fvisin) [inline main_loop_tf/main.py]: Add the formula in the comments.

File: main_loop_tf/main.py [hunk 8]
```diff
@@ -334,6 +348,9 @@ def __run(build_model):
                                              cfg.decay_steps,
                                              cfg.decay_rate,
                                              staircase=cfg.staircase)
+        elif cfg.lr_decay == 'custom':
```
Comment (fvisin) [inline main_loop_tf/main.py]: Call it STN

File: main_loop_tf/config/flow.py [hunk 9]
```diff
@@ -17,6 +17,10 @@
...
+gflags.DEFINE_bool('of_prediction', False, 'True if the running '
+                   'model performs also an optical flow prediction ')
```
Comment (fvisin) [inline main_loop_tf/config/flow.py]: Rename as `model_returns_of`

File: main_loop_tf/main.py [hunk 10]
```diff
@@ -504,18 +522,39 @@ def build_graph(placeholders, input_shape, build_model, which_set):
...
-            softmax_pred = slim.softmax(net_out)
-            tower_soft_preds.append(softmax_pred)
-            pred = tf.argmax(softmax_pred, axis=-1)
+            if not cfg.of_prediction:
+                net_out = build_model(dev_inputs, is_training)
+            else:
+                model_output = build_model(dev_inputs, is_training)
+                net_out = model_output[0]
+                of_pred = model_output[1]
+                tower_of_preds.append(of_pred)
```
Comment (fvisin) [inline main_loop_tf/main.py]: Invert if-else

File: main_loop_tf/main.py [hunk 11]
```diff
@@ -84,6 +84,12 @@ def __parse_config(argv=None):
     cfg.__dict__ = {k: el.value for (k, el) in fl.iteritems()}
     gflags.cfg = cfg
 
+    cfg.task_names = dict({'reg': 'regression',
+                           'seg': 'segmentation',
+                           'class': 'classification'})
+    assert cfg.task in cfg.task_names.values(), (
+        'Only regression/classification/segmentation are available')
```
Comment (fvisin) [inline main_loop_tf/main.py]: Remove `cfg.task_names`, use a list in the assert.

File: main_loop_tf/main.py [hunk 12]
```diff
@@ -504,18 +522,39 @@ def build_graph(placeholders, input_shape, build_model, which_set):
...
+            if not cfg.of_prediction:
+                net_out = build_model(dev_inputs, is_training)
+            else:
+                model_output = build_model(dev_inputs, is_training)
+                net_out = model_output[0]
+                of_pred = model_output[1]
+                tower_of_preds.append(of_pred)
+
+            if cfg.task in (cfg.task_names['class'],
+                            cfg.task_names['seg']):
```
Comment (fvisin) [inline main_loop_tf/main.py]: Use full names

File: main_loop_tf/main.py (hunk 11 above)
Comment (fvisin) [reply main_loop_tf/main.py]: You can add
```python
cfg.task = 'regression' if cfg.task == 'reg' else cfg.task
cfg.task = 'segmentation' if cfg.task == 'seg' else cfg.task
cfg.task = 'classification' if cfg.task == 'class' else cfg.task
```

File: main_loop_tf/main.py (hunk 12 above)
Comment (fvisin) [reply main_loop_tf/main.py]: Check for regression?

File: main_loop_tf/main.py [hunk 13]
```diff
@@ -504,18 +522,39 @@ def build_graph(placeholders, input_shape, build_model, which_set):
...
+                softmax_pred = slim.softmax(net_out)
+                tower_soft_preds.append(softmax_pred)
+                pred = tf.argmax(softmax_pred, axis=-1)
+
+                # Loss
+                if loss_fn is not sparse_softmax_cross_entropy_with_logits:
+                    # sparse_softmax_cross_entropy applies the
+                    # softmax interally
```
Comment (fvisin) [inline main_loop_tf/main.py]: internally

File: main_loop_tf/main.py [hunk 14]
```diff
@@ -504,18 +522,39 @@ def build_graph(placeholders, input_shape, build_model, which_set):
             reuse_variables = True
 
             # Model output, softmax and prediction
-            net_out = build_model(dev_inputs, is_training)
-            softmax_pred = slim.softmax(net_out)
-            tower_soft_preds.append(softmax_pred)
-            pred = tf.argmax(softmax_pred, axis=-1)
+            if not cfg.of_prediction:
+                net_out = build_model(dev_inputs, is_training)
```
Comment (fvisin) [inline main_loop_tf/main.py]: Add of_pred = None

File: main_loop_tf/main.py [hunk 15]
```diff
@@ -504,18 +522,39 @@ def build_graph(placeholders, input_shape, build_model, which_set):
...
-            if loss_fn is not sparse_softmax_cross_entropy_with_logits:
-                # sparse_softmax_cross_entropy applies the
-                # softmax internally
-                net_out = softmax_pred
-            loss = apply_loss(dev_labels, net_out, loss_fn,
+            if cfg.of_prediction and cfg.of_regularization_type is not 'None':
+                apply_loss_on = (net_out, of_pred)
+            else:
+                apply_loss_on = net_out
+            loss = apply_loss(dev_labels, apply_loss_on, loss_fn,
```
Comment (fvisin) [inline main_loop_tf/main.py]: replace with (net_out, of_pred) directly.

File: main_loop_tf/utils.py [hunk 16]
```diff
@@ -49,30 +51,53 @@ def apply_loss(labels, net_out, loss_fn, weight_decay, is_training,
...
-        print('Masking the void labels')
-        mask = tf.not_equal(labels, cfg.void_labels)
-        labels *= tf.cast(mask, 'int32')  # void_class --> 0 (random class)
-        # Train loss
-        loss = loss_fn(labels=labels,
-                       logits=tf.reshape(net_out, [-1, cfg.nclasses]))
-        mask = tf.cast(mask, 'float32')
-        loss *= mask
+    if cfg.of_prediction:
+        net_out, of_preds = net_out
```
Comment (fvisin) [inline main_loop_tf/utils.py]: Remove the if, check if of_preds is None later on.

File: main_loop_tf/main.py [hunk 17]
```diff
@@ -504,18 +522,39 @@ def build_graph(placeholders, input_shape, build_model, which_set):
...
+                net_out = build_model(dev_inputs, is_training)
+            else:
+                model_output = build_model(dev_inputs, is_training)
+                net_out = model_output[0]
+                of_pred = model_output[1]
+                tower_of_preds.append(of_pred)
+
+            if cfg.task in (cfg.task_names['class'],
+                            cfg.task_names['seg']):
+                softmax_pred = slim.softmax(net_out)
```
Comment (fvisin) [inline main_loop_tf/main.py]: rename pred_probabilities

File: main_loop_tf/main.py [hunk 18]
```diff
@@ -504,18 +522,39 @@ def build_graph(placeholders, input_shape, build_model, which_set):
...
+                tower_soft_preds.append(softmax_pred)
```
Comment (fvisin) [inline main_loop_tf/main.py]: Rename tower_pred_probabilities

File: main_loop_tf/main.py [hunk 19]
```diff
@@ -504,18 +522,39 @@ def build_graph(placeholders, input_shape, build_model, which_set):
...
+                    # sparse_softmax_cross_entropy applies the
+                    # softmax interally
+                    net_out = softmax_pred
+            else:
+                sigmoid_pred = tf.sigmoid(net_out)
+                if loss_fn is not tf.nn.sigmoid_cross_entropy_with_logits:
+                    # sigmoid_cross_entropy_with_logits applies the
+                    # sigmoid internally
+                    net_out = sigmoid_pred
+                pred = sigmoid_pred
```
Comment (fvisin) [inline main_loop_tf/main.py]: Do not do the if else on the task, but rather add a cfg.out_nonlinearity.

File: main_loop_tf/main.py [hunk 20]
```diff
@@ -676,8 +722,18 @@ def build_graph(placeholders, input_shape, build_model, which_set):
...
-        outs = [preds, softmax_preds, m_iou, per_class_iou,
-                avg_tower_loss, cm_update_op]
+        if cfg.task == cfg.task_names['reg']:
+            if cfg.of_prediction:
+                outs = [of_preds, preds, avg_tower_loss, None]
+            else:
+                outs = [preds, avg_tower_loss, None]
+        else:
+            if cfg.of_prediction:
+                outs = [of_preds, preds, softmax_preds, m_iou, per_class_iou,
```
Comment (fvisin) [inline main_loop_tf/main.py]: rename softmax_preds

File: main_loop_tf/main.py [hunk 21]
```diff
@@ -676,8 +722,18 @@ def build_graph(placeholders, input_shape, build_model, which_set):
 
         outs = [avg_tower_loss, train_ops]
     else:
-        outs = [preds, softmax_preds, m_iou, per_class_iou,
-                avg_tower_loss, cm_update_op]
+        if cfg.task == cfg.task_names['reg']:
+            if cfg.of_prediction:
+                outs = [of_preds, preds, avg_tower_loss, None]
```
Comment (fvisin) [inline main_loop_tf/main.py]: add pred_probabilities as well

Comment (fvisin) [reply main_loop_tf/main.py]: i..e, remove inner if else

File: main_loop_tf/main.py [hunk 22]
```diff
@@ -713,6 +769,8 @@ def build_graph(placeholders, input_shape, build_model, which_set):
     for _, s in enumerate(summaries[::-1]):
         summary_ops.append(tf.summary.merge(tf.get_collection_ref(key=s)))
 
+    if cfg.task == cfg.task_names['reg']:
+        reset_cm_op = None
```
Comment (fvisin) [inline main_loop_tf/main.py]: Move it where it belongs (see comment above)

File: main_loop_tf/main.py [hunk 23]
```diff
@@ -636,14 +681,15 @@ def build_graph(placeholders, input_shape, build_model, which_set):
...
-        labels, preds_flat, nclasses, mask)
+    if cfg.task == cfg.task_names['seg']:
+        # TODO Compute it for training as well (requires a dict of cms per
+        # subset + adding one_subset_per_batch to training as well)
+        # Compute the (potentially masked) mean IoU
+        mask = tf.ones_like(labels)
+        if len(cfg.void_labels):
+            mask = tf.cast(tf.less_equal(labels, nclasses), tf.int32)
+        m_iou, per_class_iou, cm_update_op, reset_cm_op = compute_mean_iou(
+            labels, preds_flat, nclasses, mask)
```
Comment (fvisin) [inline main_loop_tf/main.py:687]: add another flag cfg.compute_mean_iou

Comment (fvisin) [reply main_loop_tf/main.py:687]: else cm_update_op, reset_cm_op = None, None

Comment (fvisin) [reply main_loop_tf/main.py:687]: metrics_out = [...]

and then extend out

File: main_loop_tf/utils.py [hunk 24]
```diff
@@ -49,30 +51,53 @@ def apply_loss(labels, net_out, loss_fn, weight_decay, is_training,
...
-        # Train loss
-        loss = loss_fn(labels=labels,
-                       logits=tf.reshape(net_out, [-1, cfg.nclasses]))
+        if loss_fn is tf.losses.mean_squared_error:
+            loss = loss_fn(labels=labels,
+                           predictions=tf.reshape(net_out, [-1]))
+        elif loss_fn is tf.nn.sigmoid_cross_entropy_with_logits:
+            loss = loss_fn(labels=tf.cast(labels, cfg._FLOATX),
+                           logits=tf.reshape(net_out, [-1]))
+        # TODO: else statement
```
Comment (fvisin) [inline main_loop_tf/utils.py]: I suggest to change this as follows, if you think it makes sense:

```python
    if cfg.task in ('segmentation', 'classification') and mask_voids and len(cfg.void_labels):
        ...
    elif loss_fn is tf.losses.mean_squared_error:
        ...
    elif loss_fn is tf.nn.sigmoid_cross_entropy_with_logits:
        ...
    else
        # --> [code from L69-71]
```

File: main_loop_tf/utils.py [hunk 25]
```diff
@@ -49,30 +51,53 @@ def apply_loss(labels, net_out, loss_fn, weight_decay, is_training,
...
     if return_mean_loss:
         if mask_voids and len(cfg.void_labels):
-            return tf.reduce_sum(loss) / tf.reduce_sum(mask)
+            loss = tf.reduce_sum(loss) / tf.reduce_sum(mask)
         else:
-            return tf.reduce_mean(loss)
+            loss = tf.reduce_mean(loss)
+        if cfg.of_regularization_type is not 'None':
+            if not cfg.of_prediction:
+                raise RuntimeError('The model does not perform optical'
```
Comment (fvisin) [inline main_loop_tf/utils.py]: Add a space after 'optical'

File: main_loop_tf/utils.py [hunk 26]
```diff
@@ -49,30 +51,53 @@ def apply_loss(labels, net_out, loss_fn, weight_decay, is_training,
...
 
     # Return the mean loss (over pixels *and* batches)
     if return_mean_loss:
         if mask_voids and len(cfg.void_labels):
-            return tf.reduce_sum(loss) / tf.reduce_sum(mask)
+            loss = tf.reduce_sum(loss) / tf.reduce_sum(mask)
         else:
-            return tf.reduce_mean(loss)
+            loss = tf.reduce_mean(loss)
+        if cfg.of_regularization_type is not 'None':
```
Comment (fvisin) [inline main_loop_tf/utils.py]: Move this part in a separate fork as we discussed.

Replace it with the code to compute a secondary regression loss on the optical flow (to be added to the main loss).

File: main_loop_tf/utils.py [hunk 27]
```diff
@@ -87,6 +112,80 @@ def apply_l2_penalty(loss, weight_decay):
     return loss
 
 
+def _compute_of_local_gradients(of_preds):
```
Comment (fvisin) [inline main_loop_tf/utils.py]: Separate fork

**Review by fvisin — CHANGES_REQUESTED** [review]
I didn't check the code to visualize the OF, but since it works I suppose it's fine.

There are a few changes that I'd recommend on the rest, as we discussed yesterday. We can discuss them further if you have doubts.

**Comment by fvisin** [comment]
Closed as continued in https://github.com/fvisin/main_loop_tf/pull/19
